sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -0,0 +1,419 @@
+    this
+  }
+
+  override def size: Long = {
+    byteArray.length.toLong
+  }
+
+  override def write(src: ByteBuffer): Int = {
+    throw new UnsupportedOperationException("Read Only")
+  }
+
+  override def truncate(size: Long): SeekableByteChannel = {
+    throw new UnsupportedOperationException("Read Only")
+  }
+}
+
+/**
+ * Intermediate data structure returned from Arrow conversions
+ */
+private[sql] abstract class ArrowPayload extends Iterator[ArrowRecordBatch]
+
+/**
+ * Build a payload from existing ArrowRecordBatches
+ */
+private[sql] class ArrowStaticPayload(batches: ArrowRecordBatch*) extends ArrowPayload {
+  private val iter = batches.iterator
+  override def next(): ArrowRecordBatch = iter.next()
+  override def hasNext: Boolean = iter.hasNext
+}
+
+/**
+ * Class that wraps an Arrow RootAllocator used in conversion
+ */
+private[sql] class ArrowConverters {
+  private val _allocator = new RootAllocator(Long.MaxValue)
+
+  private[sql] def allocator: RootAllocator = _allocator
+
+  def interalRowIterToPayload(rowIter: Iterator[InternalRow], schema: StructType): ArrowPayload = {
+    val batch = ArrowConverters.internalRowIterToArrowBatch(rowIter, schema, allocator)
+    new ArrowStaticPayload(batch)
+  }
+
+  def readPayloadByteArrays(payloadByteArrays: Array[Array[Byte]]): ArrowPayload = {
+    val batches = scala.collection.mutable.ArrayBuffer.empty[ArrowRecordBatch]
+    var i = 0
+    while (i < payloadByteArrays.length) {
+      val payloadBytes = payloadByteArrays(i)
+      val in = new ByteArrayReadableSeekableByteChannel(payloadBytes)
+      val reader = new ArrowReader(in, _allocator)
+      val footer = reader.readFooter()
+      val batchBlocks = footer.getRecordBatches.asScala.toArray
+      batchBlocks.foreach(block => batches += reader.readRecordBatch(block))
+      i += 1
+    }
+    new ArrowStaticPayload(batches: _*)
+  }
+
+  def close(): Unit = {
+    _allocator.close()
+  }
+}
+
+private[sql] object ArrowConverters {
+
+  /**
+   * Map a Spark Dataset type to ArrowType.
+   */
+  private[sql] def sparkTypeToArrowType(dataType: DataType): ArrowType = {
+    dataType match {
+      case BooleanType => ArrowType.Bool.INSTANCE
+      case ShortType => new ArrowType.Int(8 * ShortType.defaultSize, true)
+      case IntegerType => new ArrowType.Int(8 * IntegerType.defaultSize, true)
+      case LongType => new ArrowType.Int(8 * LongType.defaultSize, true)
+      case FloatType => new ArrowType.FloatingPoint(FloatingPointPrecision.SINGLE)
+      case DoubleType => new ArrowType.FloatingPoint(FloatingPointPrecision.DOUBLE)
+      case ByteType => new ArrowType.Int(8, true)
+      case StringType => ArrowType.Utf8.INSTANCE
+      case BinaryType => ArrowType.Binary.INSTANCE
+      case DateType => ArrowType.Date.INSTANCE
+      case TimestampType => new ArrowType.Timestamp(TimeUnit.MILLISECOND)
+      case _ => throw new UnsupportedOperationException(s"Unsupported data type: $dataType")
```

Comment:
yes, there is a test for unsupported types here https://github.com/BryanCutler/spark/blob/wip-toPandas_with_arrow-SPARK-13534/sql/core/src/test/scala/org/apache/spark/sql/ArrowConvertersSuite.scala#L155